hal/src/dmac/transfer.rs

```diff
@@ -571,6 +595,33 @@ where
         self.complete
     }
 
+    /// Checks and clears the block transfer complete interrupt flag
+    #[inline]
+    pub fn block_transfer_interrupt(&mut self) -> bool {
+        self.chan.as_mut().check_and_clear_interrupts(
+            InterruptFlags::new().with_tcmpl(true)
+        ).tcmpl()
+    }
+
+    /// Modify a completed transfer with new `source` and `destination`, restart
+    ///
+    /// Returns false if the transfer wasn't complete, true otherwise
+    // TODO Result would be more idiomatic...
+    pub fn recycle(&mut self, mut source: S, mut destination: D) -> bool {
+        if !self.complete() {
+            return false;
+        }
+        
+        // Circular transfers won't ever complete, so never re-fill as one
+        unsafe {
+            Self::fill_descriptor(&mut source, &mut destination, false);
+        }
+
+        self.chan.as_mut().restart();
+
+        true
+    }
```

Comment:
I think this needs some additional checking to ensure it is memory safe. Consider a scenario where `S` and `D` are both `&mut [u8]`. Currently, there is nothing stopping the user from setting a 100 byte slice as source, and a 10 byte slice as destination, thus overwriting 90 bytes in memory. A length check similar to that in `new` is probably necessary. If performance is an issue, there could be an `unsafe fn recycle_unchecked` method as well.

The old source and destination buffers also need to be replaced with the new buffers in the `Transfer` struct, so that they can't be read/written while the new transfer is ongoing. Ideally, the old buffers should also be returned by the function.

A `Result` return type would probably work best here.

```suggestion
    /// Modify a completed transfer with new `source` and `destination`, 
    /// then restart transfer
    ///
    /// Returns `Err(Busy)` if the transfer wasn't complete, otherwise
    /// a tuple containing the old source and destination buffers
    // TODO: Maybe flesh out the Err variant of the return type
    pub fn recycle(&mut self, mut source: S, mut destination: D) -> Result<(S, D), ()> {
        if !self.complete() {
            return Err(());
        }
        
        // TODO: This should probably be broken out into its own function
        // in a `BufferPair` impl.
        let src_len = source.buffer_len();
        let dst_len = destination.buffer_len();

        if src_len > 1 && dst_len > 1 {
            assert_eq!(src_len, dst_len);
        }
        
        // Circular transfers won't ever complete, so never re-fill as one
        unsafe {
            Self::fill_descriptor(&mut source, &mut destination, false);
        }

        self.chan.as_mut().restart();

        Ok((
            core::mem::replace(&mut self.source, source),
            core::mem::replace(&mut self.destination, destination),
         ));
    }
```